libmtd/mtd.c

```diff
@@ -0,0 +1,466 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Operating system kernel
+ *
+ * Memory Technology Device (MTD) Interface
+ *
+ * Copyright 2022 Phoenix Systems
+ * Author: Hubert Buczynski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include "include/mtd/mtd.h"
+
+#include <errno.h>
+#include <stdlib.h>
+
+
+int mtd_erase(struct mtd_info *mtdInfo, struct erase_info *instr)
+{
+	int res;
+	storage_mtd_t *mtd = mtdInfo->storage->dev->mtd;
+
+	if (mtd->ops->erase == NULL) {
+		return -EOPNOTSUPP;
+	}
+
+	if (instr->addr >= mtdInfo->size || instr->len > (mtdInfo->size - instr->addr)) {
+		return -EINVAL;
+	}
+
+	if ((mtdInfo->flags & MTD_WRITEABLE) == 0) {
+		return -EROFS;
+	}
+
+	instr->fail_addr = MTD_FAIL_ADDR_UNKNOWN;
+	instr->state = MTD_ERASING;
+	if (instr->len == 0) {
+		instr->state = MTD_ERASE_DONE;
+		if (instr->callback != NULL) {
+			instr->callback(instr);
+		}
+
+		return EOK;
+	}
+
+	instr->state = MTD_ERASE_PENDING;
+	res = mtd->ops->erase(mtdInfo->storage, instr->addr, instr->len);
+	if (res < 0) {
+		instr->fail_addr = instr->addr;
+		instr->state = MTD_ERASE_FAILED;
+	}
+	else {
+		instr->state = MTD_ERASE_DONE;
+	}
+
+	if (instr->callback != NULL) {
+		instr->callback(instr);
+	}
```

Comment:
One exit path and code repetition less
```suggestion
	if (instr->len == 0) {
		instr->state = MTD_ERASE_DONE;
	}
	else {
		instr->state = MTD_ERASE_PENDING;
		res = mtd->ops->erase(mtdInfo->storage, instr->addr, instr->len);
		if (res < 0) {
			instr->fail_addr = instr->addr;
			instr->state = MTD_ERASE_FAILED;
		}
		else {
			instr->state = MTD_ERASE_DONE;
		}
	}

	if (instr->callback != NULL) {
		instr->callback(instr);
	}
```

Reply:
Applied Your suggestion.